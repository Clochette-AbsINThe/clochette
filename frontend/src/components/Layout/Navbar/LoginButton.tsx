import { useRouter } from 'next/router';
import { useAuthContext } from '@components/Context';
import { deleteJwtInCookie } from '@utils/auth';
import Link from 'next/link';

export default function LoginButton(): JSX.Element {
    const router = useRouter();
    const { setJwt, authenticated } = useAuthContext();

    const signOut = () => {
        setJwt(null);
        deleteJwtInCookie();
        router.push('/');
    };

    if (!authenticated) {
        return (
            <Link
                href='/login'
                className='btn-primary'>
                Se connecter
            </Link>
        );
    } else {
        return (
            <button
                onClick={signOut}
                className='btn-primary'>
                Se d√©connecter
            </button>
        );
    }
}
