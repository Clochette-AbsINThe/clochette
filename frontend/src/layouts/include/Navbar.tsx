import { DarkMode } from '@components/DarkMode';
import useWindowSize from '@hooks/useWindowSize';
import Link from 'next/link';
import { useState } from 'react';

const navitems = {
    Transaction: '/transaction',
    Stocks: '/stocks',
    Historique: '/transaction_history',
    Configuration: '/configuration'
};

export default function Navbar(): JSX.Element {
    const dimension = useWindowSize();

    const [isOpen, setIsOpen] = useState(false);

    return (
        <header className='flex flex-col'>
            <div className='flex'>
                <Link href='/'>
                    <a className='text-4xl text-green-700 font-bold p-4'>Clochette</a>
                </Link>
                {dimension.width > 1024 ? (
                    <>
                        <div className='flex p-2 justify-start flex-grow self-center h-full'>
                            {Object.entries(navitems).map(([name, link]) => (
                                <NavbarItem
                                    key={name}
                                    name={name}
                                    link={link}
                                />
                            ))}
                        </div>
                        <DarkMode />
                        <div className='justify-center self-center mr-4'>
                            <Link href='/login'>
                                <a className='btn-primary'>Login</a>
                            </Link>
                        </div>
                    </>
                ) : (
                    <>
                        <div className='flex flex-grow justify-end items-center'>
                            <div
                                id='hamburger-menu'
                                aria-label='hamburger-menu'
                                onClick={() => setIsOpen(!isOpen)}
                                className={isOpen ? 'active' : ''}>
                                <span className='bg-gray-600 dark:bg-gray-300'></span>
                                <span className='bg-gray-600 dark:bg-gray-300'></span>
                                <span className='bg-gray-600 dark:bg-gray-300'></span>
                            </div>
                        </div>
                    </>
                )}
            </div>
            {isOpen && dimension.width < 1024 && (
                <div className='relative'>
                    <div className='flex flex-grow justify-end items-center'>
                        <DarkMode />
                        <div className='justify-center self-center mr-4'>
                            <Link href='/login'>
                                <a className='btn-primary'>Login</a>
                            </Link>
                        </div>
                    </div>
                    <div className='flex flex-col w-full border-b-2 space-y-2 pb-4'>
                        {Object.entries(navitems).map(([name, link]) => (
                            <NavbarItem
                                key={name}
                                name={name}
                                link={link}
                            />
                        ))}
                    </div>
                </div>
            )}
        </header>
    );
}

function NavbarItem(props: { name: string; link: string }): JSX.Element {
    return (
        <div className='text-xl hover:underline mx-6'>
            <Link href={props.link}>
                <a>{props.name}</a>
            </Link>
        </div>
    );
}
