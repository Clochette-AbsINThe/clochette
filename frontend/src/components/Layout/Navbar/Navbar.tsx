import Link from 'next/link';
import { useState } from 'react';
import dynamic from 'next/dynamic';
import LoginButton from '@components/Layout/Navbar/LoginButton';
import { useAuthContext } from '@components/Context';
import { getRedirectUrlEncoded } from '@utils/utils';

const navitems = {
    Transaction: '/transaction',
    Stock: '/stock',
    Historique: '/transaction_history',
    Configuration: '/configuration'
};

// Import Dark mode dynamically to avoid SSR issues
const DarkMode = dynamic(() => import('@components/Layout/Navbar/DarkMode').then((mod) => mod.DarkMode), { ssr: false });

export default function Navbar(): JSX.Element {
    const [isOpen, setIsOpen] = useState(false);

    return (
        <header className='flex flex-col lg:flex-row'>
            <div className='flex justify-between items-center'>
                <Link
                    href='/'
                    className='text-4xl text-green-700 font-bold p-4'>
                    Clochette
                </Link>
                <div
                    id='hamburger-menu'
                    aria-label='hamburger-menu'
                    onClick={() => setIsOpen(!isOpen)}
                    className={(isOpen ? 'active ' : '') + 'lg:hidden'}>
                    <span className='bg-gray-600 dark:bg-gray-300'></span>
                    <span className='bg-gray-600 dark:bg-gray-300'></span>
                    <span className='bg-gray-600 dark:bg-gray-300'></span>
                </div>
            </div>
            <div className={(!isOpen ? 'hidden ' : '') + 'lg:flex flex-col lg:flex-row items-center flex-grow'}>
                <div className={'flex p-2 justify-start flex-grow self-start lg:self-center h-full flex-col lg:flex-row pb-4 space-y-2 lg:space-y-0 lg:items-end lg:mb-2'}>
                    {Object.entries(navitems).map(([name, link]) => (
                        <NavbarItem
                            key={name}
                            name={name}
                            link={link}
                        />
                    ))}
                </div>
                <div className='flex flex-row justify-end'>
                    <DarkMode />
                    <div className='justify-center self-center mr-4'>
                        <LoginButton />
                    </div>
                </div>
            </div>
        </header>
    );
}

function NavbarItem(props: { name: string; link: string }): JSX.Element {
    const { authenticated } = useAuthContext();
    const link = authenticated ? props.link : `/login?redirect=${getRedirectUrlEncoded(props.link)}`;
    return (
        <div className='text-xl hover:underline mx-6 w-max'>
            <Link href={link}>{props.name}</Link>
        </div>
    );
}
