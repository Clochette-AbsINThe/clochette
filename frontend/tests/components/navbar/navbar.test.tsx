import { fireEvent, render, screen } from '@testing-library/react';

import Navbar from '@/components/navbar/navbar';

describe('Navbar', () => {
  it('should match snapshot', () => {
    const { container } = render(<Navbar />);
    expect(container).toMatchSnapshot();
  });

  it('should open the mobile menu when clicking on the hamburger button', () => {
    window.innerWidth = 500;
    render(<Navbar />);
    fireEvent.click(screen.getByLabelText('hamburger-menu'));
    const login = screen.queryByText('Se d√©connecter');
    expect(login).not.toBeNull();
  });
});
