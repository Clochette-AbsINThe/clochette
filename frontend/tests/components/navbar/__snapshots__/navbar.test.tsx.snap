// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Navbar > should match snapshot 1`] = `
<div>
  <header
    class="flex flex-col lg:flex-row"
  >
    <div
      class="flex justify-between items-center"
    >
      <a
        class="text-4xl text-green-700 font-bold p-2"
        href="/"
      >
        Clochette
      </a>
      <button
        aria-label="hamburger-menu"
        class="lg:hidden"
        id="hamburger-menu"
      >
        <span
          class="bg-gray-600 dark:bg-gray-300"
        />
        <span
          class="bg-gray-600 dark:bg-gray-300"
        />
        <span
          class="bg-gray-600 dark:bg-gray-300"
        />
      </button>
    </div>
    <div
      class="hidden lg:flex flex-col lg:flex-row items-center flex-grow"
    >
      <div
        class="flex p-2 justify-start flex-grow self-start lg:self-center h-full flex-col lg:flex-row pb-4 space-y-2 lg:space-y-0 lg:items-end lg:mb-2"
      >
        <a
          class="text-xl hover:underline mx-4 w-max xl:mx-6"
          href="/transaction/vente"
        >
          Transaction
        </a>
        <a
          class="text-xl hover:underline mx-4 w-max xl:mx-6"
          href="/stock"
        >
          Stock
        </a>
        <a
          class="text-xl hover:underline mx-4 w-max xl:mx-6"
          href="/transaction-history"
        >
          Historique
        </a>
        <a
          class="text-xl hover:underline mx-4 w-max xl:mx-6"
          href="/configuration"
        >
          Configuration
        </a>
      </div>
      <div
        class="flex flex-row justify-end flex-wrap lg:flex-nowrap"
      >
        <div
          class="justify-center self-center mr-4"
        >
          <div
            class="flex"
          >
            <button
              class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
            >
              Se d√©connecter
            </button>
            <a
              class="ml-4"
              href="/account"
            >
              <button
                class="inline-flex items-center justify-center rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
              >
                <svg
                  class="h-5 w-5"
                  fill="none"
                  height="15"
                  viewBox="0 0 15 15"
                  width="15"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M7.5 0.875C5.49797 0.875 3.875 2.49797 3.875 4.5C3.875 6.15288 4.98124 7.54738 6.49373 7.98351C5.2997 8.12901 4.27557 8.55134 3.50407 9.31167C2.52216 10.2794 2.02502 11.72 2.02502 13.5999C2.02502 13.8623 2.23769 14.0749 2.50002 14.0749C2.76236 14.0749 2.97502 13.8623 2.97502 13.5999C2.97502 11.8799 3.42786 10.7206 4.17091 9.9883C4.91536 9.25463 6.02674 8.87499 7.49995 8.87499C8.97317 8.87499 10.0846 9.25463 10.8291 9.98831C11.5721 10.7206 12.025 11.8799 12.025 13.5999C12.025 13.8623 12.2376 14.0749 12.5 14.0749C12.7623 14.075 12.975 13.8623 12.975 13.6C12.975 11.72 12.4778 10.2794 11.4959 9.31166C10.7244 8.55135 9.70025 8.12903 8.50625 7.98352C10.0187 7.5474 11.125 6.15289 11.125 4.5C11.125 2.49797 9.50203 0.875 7.5 0.875ZM4.825 4.5C4.825 3.02264 6.02264 1.825 7.5 1.825C8.97736 1.825 10.175 3.02264 10.175 4.5C10.175 5.97736 8.97736 7.175 7.5 7.175C6.02264 7.175 4.825 5.97736 4.825 4.5Z"
                    fill="currentColor"
                    fill-rule="evenodd"
                  />
                </svg>
              </button>
            </a>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;
